source/mongodb-handshake/handshake.rst

```diff
@@ -224,6 +224,107 @@ Example::
         - "Oracle JVM EE 9.1.1"
 
 
+--------------------------
+Speculative Authentication
+--------------------------
+
+:since: 4.4
+
+The ``isMaster`` handshake supports a new argument, ``speculativeAuthenticate``,
+provided as a BSON object. Clients specifying this argument to ``isMaster`` will
+speculatively include the first command of an authentication handshake.
+This command may be provided to the server in parallel with any standard request for
+supported authentication mechanisms (i.e. ``saslSupportedMechs``). This would permit
```

Comment:
It's not clear to me what the exact interaction with saslSupportedMechs is. This is pymongo's current code that discovers whether we should authenticate with SCRAM-SHA-256 or SCRAM-SHA-1 (only used when the user does not specify an authMechanism):
```python
        cmd = SON([
            ('ismaster', 1),
            ('saslSupportedMechs', source + '.' + credentials.username)])
        mechs = sock_info.command(
            source, cmd, publish_events=False).get('saslSupportedMechs', [])
        if 'SCRAM-SHA-256' in mechs:
            return _authenticate_scram(credentials, sock_info, 'SCRAM-SHA-256')
        else:
            return _authenticate_scram(credentials, sock_info, 'SCRAM-SHA-1')
```

Are we supposed to send speculativeAuthenticate? If so, do we choose SCRAM-SHA-1 or SCRAM-SHA-256? What happens if we choose the wrong mechanism?

Reply:
Sending `speculativeAutheticate` is optional. It reduces the number of round trips by 1. The paragraph starting on line 261 states that the SCRAM-SHA-256 mechanism should be used if no mechanism was specified. If the mechanism is not supported, the server will not send a `speculativeAuthenticate` reply in the `isMaster` response, and the standard authentication handshake must be performed.

Reply:
Thanks! I made a small suggestion in that paragraph.